pyo3-build-config/src/util.rs

```diff
@@ -0,0 +1,59 @@
+//! Additional functionality enabled by crate features
+
+#![allow(clippy::unnecessary_wraps)]
+
+use super::Triple;
+use crate::errors::Result;
+
+/// Generates `python3.dll` import library for Windows targets.
+///
+/// Places the import library into the build script output directory
+/// and returns the full library directory path.
+///
+/// Does nothing if the target OS is not Windows.
+#[cfg(feature = "python3-dll-a")]
+pub(super) fn generate_abi3_import_lib(target: &Triple) -> Result<Option<String>> {
+    use super::{Architecture, OperatingSystem};
+    use crate::errors::Context;
+
+    if target.operating_system != OperatingSystem::Windows {
+        return Ok(None);
+    }
+
+    let out_dir = std::env::var_os("OUT_DIR")
+        .expect("generate_abi3_import_lib() must be called from a build script");
+
+    // Put the import library into the build script output directory.
+    let mut out_lib_dir = std::path::PathBuf::from(out_dir);
+    out_lib_dir.push("lib");
+
+    // Convert `Architecture` to rustc `target_arch` format.
+    let arch = match target.architecture {
+        // i686, i586, etc.
+        Architecture::X86_32(_) => "x86".to_string(),
+        other => other.to_string(),
+    };
+
+    let env = target.environment.to_string();
+
+    python3_dll_a::generate_implib_for_target(&out_lib_dir, &arch, &env)
+        .context("failed to generate python3.dll import library")?;
+
+    let out_lib_dir_string = out_lib_dir
+        .to_str()
+        .ok_or("build directory is not a valid UTF-8 string")?
+        .to_owned();
+
+    Ok(Some(out_lib_dir_string))
+}
+
+/// Generates `python3.dll` import library for Windows targets.
+///
+/// Places the import library into the build script output directory
+/// and returns the full library directory path.
+///
+/// Does nothing if the target OS is not Windows.
+#[cfg(not(feature = "python3-dll-a"))]
```

Comment:
I think it would be nicer to make the invocation dependent on the feature in this case as right now, `_impls.rs` looks like something (maybe some fallback logic) might happen when actually this is a no-op. (If the module was feature-specific, the `mod` statement itself could carry the `cfg` directive (in addition to the invocation).)

Reply:
This feature uses an external crate that is an optional dependency. I can add an `if cfg!() {}` statement to `_impl.rs`, but even then I'd still need to define two versions of `generate_abi3_import_lib()`.

Reply:
I was thinking of something like

```rust
#[cfg(feature = "python3-dll-a")]
#[path = "util.rs"]
mod python3_dll_a;
...
pub fn make_interpreter_config() -> Result<InterpreterConfig> {
...
    // Auto generate python3.dll import libraries for Windows targets.
    #[cfg(feature = "python3-dll-a")]
    interpreter_config.lib_dir = python3_dll_a::generate_abi3_import_lib(&host)?;
...
}
```

in `_impl.rs` and then no `cfg`-directives at all in `python3_dll_a.rs`.

Reply:
Do `#[cfg(...)]` directives work at the statement level? I didn't know that...

Reply:
I think they are still experiment on the expression level, c.f. https://github.com/rust-lang/rust/issues/15701 but that can usually be worked around by adding a block, i.e. both

```rust
...
#[cfg(feature = "python3-dll-a")]
{
    // Auto generate python3.dll import libraries for Windows targets.
    interpreter_config.lib_dir = python3_dll_a::generate_abi3_import_lib(&host)?;
}
...
```

and

```rust
...
#[cfg(feature = "python3-dll-a")]
if abi3 && lib_dir.is_none() {
    // Auto generate python3.dll import libraries for Windows targets.
    lib_dir = python3_dll_a::generate_abi3_import_lib(&host)?;
}
...
```

should work.

(This can provide dead-code and unneeded-mut noise. So it isn't totally clear that it is the better. My primary aim here is to make it visible at the call site, that the operation is optional depending on whether the dependency is present.)

Reply:
Thanks for the feedback! I'll try to apply these suggestions and submit the new patch tomorrow.

I can't name the submodule `python3_dll_a` because it's also the external crate name, so I'll go with `abi3_import_lib` or something.